### PR TITLE
fix(enterchat): fixing statstring buffer overwrite of product

### PR DESCRIPTION
**Pull request by richard-pianka**

### CODE REVIEW

**Comment by richard-pianka** [comment]
@carlbennett ready

**Comment by carlbennett** [comment]
This seems bugged and not ready yet. The product id is messed up in both chat gateway and the game protocols.

Chat Gateway: `1001 USER pianka 0000 [aip,]`

image

**Comment by richard-pianka** [comment]
per my discord comment:

> relatively sure i figured it out: diablo 2 appears to be the only game that we need to fix the statstring for.  in particular though, bnet ignores the user-provided statstring when its an open character.  further, the client crashes with statstrings that are only PRODRealm,char-name for some reason.  regardless, bnet doesn't show users anything about the open character's name or items to render.  i'll update the PR shortly.

this doesn't support realm character statstrings yet, but we also don't have realm servers regardless

File: src/Atlasd/Battlenet/Protocols/Game/Messages/SID_ENTERCHAT.cs [hunk 1]
```diff
@@ -72,9 +66,14 @@ public override bool Invoke(MessageContext context)
...
 
+                        if (Product.IsDiabloII(gameState.Product))
+                        {
+                            statstring = Product.ToByteArray(gameState.Product);
+                        }
+
                         // Do not use client-provided statstring if config.battlenet.emulation.statstring_updates is not enabled for this product.
                         // Blizzard servers allowed statstring updates for Diablo, Diablo II (changing characters), Warcraft III (changing icons), and Shareware variants.
-                        if (!GameState.CanStatstringUpdate(gameState.Product)) statstring = gameState.GenerateStatstring();
+                        if (!GameState.CanStatstringUpdate(gameState.Product) || statstring.Length == 0) statstring = gameState.GenerateStatstring();
```
Comment (carlbennett) [inline src/Atlasd/Battlenet/Protocols/Game/Messages/SID_ENTERCHAT.cs:76]: I believe this code causes Diablo II clients to always by interpreted/seen as open realm, as it actively resets the info the Diablo II client provided to being simply the D2DV/D2XP product code before finally calling the statstring generator.

**Review by carlbennett — DISMISSED** [review]
Tentatively approving so the code can be merged. Diablo II may need more eyes than I care to test for at this time.

**Review by carlbennett — APPROVED** [review]
Re-approving

**Pull request merged**